### PR TITLE
Feature #16812: Allow slicing arrays with ArithmeticSequence

### PR DESCRIPTION
**Pull request by mrkn**
TODO:
- [ ] update documentation after we decide the spec

### CODE REVIEW

**Comment by mrkn** [comment]
@nobu Could you check whether or not `ary_make_partial_step` correctly handles array’s memory?

File: array.c [hunk 1]
```diff
@@ -1201,6 +1201,52 @@ ary_make_partial(VALUE ary, VALUE klass, long offset, long len)
...
+    assert(offset+len <= RARRAY_LEN(ary));
+    assert(step != 0);
+
+    const long orig_len = len;
+
+    if ((step > 0 && step >= len) || (step < 0 && (step < -len))) {
+        VALUE result = ary_new(klass, 1);
+        VALUE *ptr = (VALUE *)ARY_EMBED_PTR(result);
+        *ptr = RARRAY_AREF(ary, offset);
+        ARY_SET_EMBED_LEN(result, len);
```
Comment (nobu) [inline array.c:1218]: Is `len` embeddable here?

File: array.c [hunk 2]
```diff
@@ -1201,6 +1201,52 @@ ary_make_partial(VALUE ary, VALUE klass, long offset, long len)
...
+    long ustep = (step < 0) ? -step : step;
+    len = (len + ustep - 1) / ustep;
+
+    long i;
+    long j = offset + ((step > 0) ? 0 : (orig_len - 1));
+    VALUE result = ary_new(klass, len);
+    if (len <= RARRAY_EMBED_LEN_MAX) {
+        VALUE *ptr = (VALUE *)ARY_EMBED_PTR(result);
+        for (i = 0; i < len; ++i) {
+            *ptr++ = RARRAY_AREF(ary, j);
```
Comment (nobu) [inline array.c:1231]: Isn’t WP needed here?

File: array.c [hunk 3]
```diff
@@ -1201,6 +1201,52 @@ ary_make_partial(VALUE ary, VALUE klass, long offset, long len)
...
+            j += step;
+        }
+        ARY_SET_EMBED_LEN(result, len);
+    }
+    else {
+        const VALUE *values = RARRAY_CONST_PTR_TRANSIENT(ary);
```
Comment (nobu) [inline array.c:1237]: Maybe unused?

File: range.c [hunk 4]
```diff
@@ -1282,50 +1282,58 @@ rb_range_values(VALUE range, VALUE *begp, VALUE *endp, int *exclp)
...
     if (NIL_P(e)) excl = 0;
     origbeg = beg;
     origend = end;
     if (beg < 0) {
-	beg += len;
-	if (beg < 0)
-	    goto out_of_range;
+        beg += len;
+        if (beg < 0)
+            goto out_of_range;
```
Comment (nobu) [inline range.c:1298]: What changes here?

Comment (mrkn) [reply range.c:1298]: Only expanding tabs.

File: array.c (hunk 2 above)
Comment (nobu) [reply array.c:1231]: Sorry, WB (write-barrier).

**Comment by mrkn** [comment]
@nobu I fixed `ary_make_partial_step`. Could you review it again?

**Pull request merged**